Added AssetGrid/Timeline to Map

RBedbur opened the pull request:
## Description

Added an AssetGrid/Timeline to map page.
By default the grid is closed to avoid loading a large set of timebuckets and images.
When Grid is opened it updates when map is zoomed or relocated.

Implements feature #4519 


Code partly from discussion #11893 


Did not add an option to mapsettings. The grid is always available (but closed).

Next steps:
Show App Bar, when images are selected in grid. Right now nothing happens on selecting images.
Show images from the same timeframe as the current images shown in grid. This could be useful when shooting with a phone (geolocation is set) and a mirrorless camera (without geolocation).

## How Has This Been Tested?

Manually tested with 6000 geolocated images and around 1300 images per location.

<details><summary><h2>Screenshots (if appropriate)</h2></summary>

<!-- Images go below this line. -->
<img width="1123" alt="image" src="https://github.com/user-attachments/assets/a043f401-ec61-4ca6-b6fd-7e3f5a462606" />
<img width="1112" alt="image" src="https://github.com/user-attachments/assets/09224d68-beda-43db-a70e-27aa08bfaa20" />
</details>

## API Changes

api/timeline/buckets and api/timeline/bucket accept coordinates as x1, x2, y1, y2 parameters
updated TimeBucketDto



## Checklist:

- [X ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation if applicable
- [ X] I have no unrelated changes in the PR.
- [X ] I have confirmed that any new dependencies are strictly necessary. (no new dependencies)
- [ ] I have written tests for new code (if applicable)
- [X ] I have followed naming conventions/patterns in the surrounding code
- [ X] All code in `src/services` uses repositories implementations for database calls, filesystem operations, etc. (no new code)
- [ X] All code in `src/repositories/` is pretty basic/simple and does not have any immich specific logic (that belongs in `src/services`) (no new code)

RBedbur committed:
Showing timeline on map

RBedbur committed:
Toggling Timeline

RBedbur committed:
Showing Timeline in Fullscreen

RBedbur committed:
Merge pull request #1 from RBedbur/feat/discussion-4519

Feat/discussion 4519

RBedbur committed:
Merge branch 'immich-app:main' into main

RBedbur committed:
Refactores map control and map/page control
Moved AssetGrid from map to map/page

RBedbur committed:
Merge branch 'immich-app:main' into main

YarosMallorca commented:
Does this truly need server modifications?
Just wondering because such a feature is present on the mobile app already.

jrasm91 commented:
The mobile app can query the local database.

danieldietzler commented on web/src/lib/constants.ts:
```diff
@@ -70,10 +70,51 @@ export const dateFormats = {
+  assetGridOptions?: AssetGridOptions;
+}
+
+export const createAssetGridOptionsFromArray = (enumArray: string[]): AssetGridOptions => {
```
`enumArray` is a very nondescriptive name

danieldietzler commented on web/src/lib/utils.ts:
```diff
@@ -343,3 +343,13 @@ export const withError = async <T>(fn: () => Promise<T>): Promise<[undefined, T]
 // eslint-disable-next-line unicorn/prefer-code-point
 export const decodeBase64 = (data: string) => Uint8Array.from(atob(data), (c) => c.charCodeAt(0));
+
+export const parseNumberOrUndefined = (value: string | null): number | undefined => {
+  return value !== null && !Number.isNaN(Number(value)) ? Number(value) : undefined;
+};
```
I don't think we need this helper here.

danieldietzler commented on web/src/routes/(user)/map/[[photos=photos]]/[[assetId=id]]/+page.svelte:
```diff
@@ -18,6 +18,22 @@
+  import { mdiArrowDown, mdiArrowUp } from '@mdi/js';
+  import Button from '$lib/components/elements/buttons/button.svelte';
+  import Icon from '$lib/components/elements/icon.svelte';
+  import maplibregl from 'maplibre-gl';
```
That this bleeds out into other pages is kind of unfortunate and probably indicates that the architecture isn't ideal.

danieldietzler commented on web/src/routes/(user)/map/[[photos=photos]]/[[assetId=id]]/+page.svelte:
```diff
@@ -130,12 +186,97 @@
+      lng_e = -180;
+      lng_w = 180;
+    } else if (Math.abs(bounds._sw.lng) > 180) {
+      lng_e = bounds._sw.lng + 360;
```
Where did you get from that you need to add 360 if `|lng| > 180`? Is that per the maplibre docs?

danieldietzler commented on web/src/routes/(user)/map/[[photos=photos]]/[[assetId=id]]/+page.svelte:
```diff
@@ -130,12 +186,97 @@
   <UserPageLayout title={data.meta.title}>
-    <div class="isolate h-full w-full">
-      <Map hash bind:mapMarkers bind:showSettingsModal onSelect={onViewAssets} />
+    <div class="isolate h-full w-full" id="map-container" bind:this={mapContainer}>
```
I don't think we need to keep it inside the map container. In fact, it feels rather odd doing that to me

RBedbur replied:
This is used to resize the map to fullscreen when pressing the fullscreen button on the map. Then also the asset grid is visible. Otherwise only the map would be resized to fullscreen and the asset grid is hidden.

danieldietzler replied:
I don't understand. Why do you need to have the container in order to determine if some button was pressed?

RBedbur replied:
When pressing the Fullscreen Button on the map, it will resize the map only. The asset grid below will not be visible.
By passing in the mapContainer to the FullScreenControl (), everything inside the container (map, toggle button and assetgrid) is part of the fullscreen.

danieldietzler commented on web/src/routes/(user)/people/+page.svelte:
```diff
@@ -243,7 +243,7 @@
   const handleMergePeople = async (detail: PersonResponseDto) => {
     await goto(
-      `${AppRoute.PEOPLE}/${detail.id}?${QueryParameter.ACTION}=${ActionQueryParameterValue.MERGE}&${QueryParameter.PREVIOUS_ROUTE}=${AppRoute.PEOPLE}`,
+      `${AppRoute.PEOPLE}/${detail.id}?${QueryParameter.ACTION}=${ActionQueryParameterValue.MERGE}&${QueryParameter.PREVIOUS_ROUTE}=${encodeURIComponent(AppRoute.PEOPLE)}`,
```
Again, unrelated and not necessary (afaik)

danieldietzler commented on web/src/routes/(user)/people/[personId]/[[photos=photos]]/[[assetId=id]]/+page.svelte:
```diff
@@ -112,10 +112,10 @@
     const getPreviousRoute = $page.url.searchParams.get(QueryParameter.PREVIOUS_ROUTE);
-    if (getPreviousRoute && !isExternalUrl(getPreviousRoute)) {
-      previousRoute = getPreviousRoute;
+    if (getPreviousRoute && !isExternalUrl(decodeURIComponent(getPreviousRoute))) {
```
Same here.

danieldietzler commented on web/src/routes/(user)/photos/[[assetId=id]]/+page.svelte:
```diff
@@ -32,8 +32,17 @@
   import { t } from 'svelte-i18n';
 
+  // TODO: when getTimebuckets support withArchived
+  const isArchived = false;
+  const withStacked = true;
+
   let { isViewing: showAssetViewer } = assetViewingStore;
-  const assetStore = new AssetStore({ isArchived: false, withStacked: true, withPartners: true });
+  const assetStore = new AssetStore({
+    isArchived,
+    withStacked,
+    withPartners: false,
+    isFavorite: false,
+  });
```
Why this change?

danieldietzler reviewed (commented):
Thanks for the PR. Right now it appears quite complex to me (which in parts is probably also due to intrinsic complexity), with some random and unrelated changes. Also, please revert the unrelated formatting changes in the i18n files and the vscode settings.

mertalev commented on server/src/dtos/time-bucket.dto.ts:
```diff
@@ -41,6 +43,38 @@ export class TimeBucketDto {
+  @IsNumber()
+  @Min(-180)
+  @Max(180)
+  @Type(() => Number)
```
`@IsNumber` might be needed to reject string numbers

danieldietzler replied:
https://github.com/typestack/class-validator/blob/develop/src/decorator/number/Max.ts#L9-L11

mertalev replied:
Oh nice. I figured it was just a `<` / `>` comparison

RBedbur replied:
Without `@Type(() => Number)` the input will not be casted from string to number.

mertalev replied:
I prefer rejecting strings over casting them, personally. But this is up to preference.

RBedbur replied:
The input comes from QueryString which is always string if not casted. Compare other Dtos from Query string, where you already have Type conversions, e.g. [RandomAssetsDto](https://github.com/immich-app/immich/blob/6c5f99c47aa4f2a9e56a0041cce7ae6f4ea75a94/server/src/dtos/asset.dto.ts#L76)

RBedbur committed:
Removed unrelated changes
Refactored code

danieldietzler commented on web/src/lib/components/shared-components/map/map.svelte:
```diff
@@ -145,21 +160,22 @@
       {#if !simplified}
         <GeolocateControl position="top-left" />
-        <FullscreenControl position="top-left" />
+        <FullscreenControl position="top-left" container={mapContainer} />
```
Gotcha. I was hoping it'd expose some callback function, but doesn't look like it. Kind of ugly though :/

jrasm91 commented:
This type of query won't perform very good on a lot of assets. Zack and I talked about adding a polygon based index, similar to the other geocoding stuff that we already have.

RBedbur commented:
> This type of query won't perform very good on a lot of assets. Zack and I talked about adding a polygon based index, similar to the other geocoding stuff that we already have.

Ok. That means?
I just refactored / cleaned the code and implemented all comments from @danieldietzler.

Cann you point me to some code, where you implemented the polygon based index?

RBedbur committed:
Moved bounds normjalization to map

RBedbur committed:
Refactored Property Decorators in TimeBucketDto

RBedbur commented:
Thanks for the feedback.
The last thing I heard about were thoughts about performance, which is understandable as there is a lot of data loaded from the server to place the markers on the map. 
@danieldietzler any idea how to go on?

danieldietzler commented:
As jrasm already said the key here is probably to fetch assets that are currently in frame. I don't really have experience with polygon/geospatial indices though. That's probably something @zackpollard could chime in on.

SenderLibs commented:
Is this still being worked on?
